File: ga/asg1/graph_c/implementations/vertex.cpp
#include <algorithm>
#include "../interfaces/iterator.h"
#include "../interfaces/vertex.h"


Vertex::Vertex(int label) {
  this->label = label;
}


void Vertex::remove_inbound(int label) {
  auto it = std::find(this->inbound.begin(), this->inbound.end(), label);

  if(it == this->inbound.end())
    return;

  this->inbound.erase(it);
}


void Vertex::remove_outbound(int label) {
  auto it = std::find(this->outbound.begin(), this->outbound.end(), label);

  if(it == this->outbound.end())
    return;

  this->outbound.erase(it);
}


int Vertex::get_degree_in() const {
  return (int)this->inbound.size();
}


int Vertex::get_degree_out() const {
  return (int)this->outbound.size();
}


bool Vertex::add_inbound(int label) {
  auto it = std::find(this->inbound.begin(), this->inbound.end(), label);
  if(it == this->inbound.end()) {
    this->inbound.push_back(label);
    return true;
  }
  return false;
}


bool Vertex::add_outbound(int label) {
  auto it = std::find(this->outbound.begin(), this->outbound.end(), label);
  if(it == this->outbound.end()) {
    this->outbound.push_back(label);
    return true;
  }
  return false;
};


Iterator<int> Vertex::get_inbound_edges_iterator() const {
  return Iterator<int>(this->inbound);
}


Iterator<int> Vertex::get_outbound_edges_iterator() const {
  return Iterator<int>(this->outbound);
}

File: ga/asg1/graph_python/Vertex.py
from typing import TypeVar
from Iterator import Iterator


T = TypeVar('T')


class Vertex:

    def __init__(self, label: T):
        self.label = label
        self.__inbound = []
        self.__outbound = []


    def remove_inbound(self, label: T):
        try:
            idx = self.inbound.index(label)
            self.pop(idx)
        except ValueError:
            pass


    def remove_outbound(self, label: T):
        try:
            idx = self.outbound.index(label)
            self.pop(idx)
        except ValueError:
            pass


    def get_degree_in(self) -> int:
        return len(self.inbound)


    def get_degree_out(self) -> int:
        return len(self.outbound)


    def add_inbound(self, label: T) -> bool:
        try:
            idx = self.inbound.index(label)
            return False
        except ValueError:
            self.inbound.append(label)
            return True


    def add_outbound(self, label: T) -> bool:
        try:
            idx = self.outbound.index(label)
            return False
        except ValueError:
            self.outbound.append(label)
            return True


    def get_inbound_edges_iterator(self) -> Iterator:
        return Iterator(self.inbound)


    def get_outbound_edges_iterator(self) -> Iterator:
        return Iterator(self.outbound)
